File: IRC Bot C++/IRC Bot/IRC Bot.cpp
#include "stdafx.h"

#pragma comment(lib, "Ws2_32.lib")

class IrcBot
{
	public:
		IrcBot(char* real, char* user, char* host, char* nick, char* server, char* channel, char* port);
		char* getRealName() { return realName; }
		char* getUserName() { return userName; }
		char* getNickName() { return nickName; }
		char* getHostName() { return hostName; }
		char* getServer() { return serverName; }
		char* getChannel() { return channelName; }
		char* getPort() { return port; }

		std::vector<char*> names_to_op =
		{
			"Seanharrs", "SlayerSean",
			"AusBotPython", "Aus_Bot_Python",
			"AusBotCSharp", "Aus_Bot_CSharp",
			"AusBotFSharp", "Aus_Bot_FSharp"
		};

	private:
		char* realName;
		char* userName;
		char* nickName;
		char* hostName;
		char* serverName;
		char* channelName;
		char* port;
};

IrcBot::IrcBot(char* real, char* user, char* host, char* nick, char* server, char* channel, char* port = "6667")
{
	realName = real;
	userName = user;
	hostName = host;
	nickName = nick;
	serverName = server;
	channelName = channel;
	this->port = port;
}

class Program
{
	public:
		void main();

		enum State { DO_NOTHING, CONTINUE, RETURN };

	private:
		std::vector<std::string> explode(const std::string& str, const char& delim1, const char& delim2);
		State checkServerMessages(std::string line);
		State checkUserMessages(std::string line);
		void sendData(char* command, char* param);
		void openConnection();
		void disconnect();
		void join();

		struct addrinfo *result = nullptr, *ptr = nullptr, hints;

		SOCKET irc = INVALID_SOCKET;
		IrcBot* bot;
		std::string unusedData;
};

void Program::sendData(char* command, char* param)
{
	char message[1024];
	sprintf_s(message, "%s %s", command, param);
	std::cout << message << std::endl;
	send(irc, message, int(strlen(message)), 0);
}

void Program::openConnection()
{
	WSADATA wsaData;
	int res = WSAStartup(MAKEWORD(2, 2), &wsaData);
	if (res != 0) throw std::exception("Startup failed!");

	ZeroMemory(&hints, sizeof(hints));
	hints.ai_family = AF_UNSPEC;
	hints.ai_socktype = SOCK_STREAM;
	hints.ai_protocol = IPPROTO_TCP;
	res = getaddrinfo("irc.quakenet.org", bot->getPort(), &hints, &result);
	if (res != 0) throw std::exception("Get address info failed!");

	ptr = result;
	irc = socket(ptr->ai_family, ptr->ai_socktype, ptr->ai_protocol);
	if (irc == INVALID_SOCKET)
	{
		freeaddrinfo(result);
		throw std::exception("Failed to create socket: " + WSAGetLastError());
	}
	res = connect(irc, ptr->ai_addr, int(ptr->ai_addrlen));
	if (res == SOCKET_ERROR)
	{
		closesocket(irc);
		irc = INVALID_SOCKET;
		freeaddrinfo(result);
		throw std::exception("Socket connection failed!");
	}
	sendData("NICK", bot->getNickName());

	char param[1024] = "";
	sprintf_s(param, "%s %s %s :%s", bot->getUserName(), bot->getHostName(), bot->getServer(), bot->getRealName());
	sendData("USER", param);
}

void Program::disconnect()
{
	//shutdown(irc, SD_SEND);
	closesocket(irc);
	WSACleanup();
}

void Program::join() { sendData("JOIN", bot->getChannel()); }

Program::State Program::checkServerMessages(std::string line)
{
	std::smatch match;
	if (regex_search(line, match, std::regex("PING :(\S*)")))
	{
		for (auto x : match) std::cout << x;
		std::cout << "HEY!";
		//sendData("PONG", )
	}
	else if (regex_search(line, match, std::regex(":([^!]*)!")))
	{
		for (auto x : match) std::cout << x;
		std::cout << "HEY!";
		//sendData("JOIN", )
	}
	else if (regex_search(line, match, std::regex("[+-]o (\S*)")))
	{
		for (auto x : match) std::cout << x;
		std::cout << "HEY!";
		//sendData("MODE", )
	}
	else return DO_NOTHING;

	return CONTINUE;
}

Program::State Program::checkUserMessages(std::string line)
{
	std::smatch match;
	if (regex_search(line, match, std::regex(".*?\.botquit.*?"))) disconnect();
	else if (regex_search(line, match, std::regex(":([^!]*)!.*?:.get ([\S ]*)")))
	{
		for (auto x : match) std::cout << x;
		std::cout << "HEY!";
		//sendData("PRIVMSG", )
	}
	else return DO_NOTHING;

	return CONTINUE;
}

std::vector<std::string> Program::explode(const std::string& message, const char& delim1, const char& delim2)
{
	std::vector<std::string> data;
	for (std::string::const_iterator iter = message.begin(); iter != message.end(); ++iter)
	{
		if ((*iter == delim1 || *iter == delim2) && !unusedData.empty())
		{
			data.push_back(unusedData);
			unusedData.clear();
		}
		else if(*iter != delim1 && *iter != delim2) unusedData += *iter;
	}
	return data;
}

void Program::main()
{
	bot = new IrcBot("Aus Bot CPlusPlus v0", "AUS_Bot_CPlusPlus", "AusBotHosting", "AusBotCPlusPlus", "irc.quakenet.org", "#ausbot_test");
	try
	{
		openConnection();
		int count = 0;
		int failCount = 0;
		std::vector<char> recvBuffer(1024);
		while (true)
		{
			int res = recv(irc, recvBuffer.data(), recvBuffer.size(), 0);
			if (res > 0)
			{
				std::vector<std::string> lines = explode(recvBuffer.data(), '\r', '\n');

				for (size_t i = 0; i < lines.size(); ++i)
				{
					std::string line = lines[i];
					std::cout << lines[i] << std::endl;
					++count;
					if (count == 4)
					{
						join();
						continue;
					}
					switch (checkServerMessages(lines[i]))
					{
						case RETURN: return;
						case CONTINUE: continue;
						case DO_NOTHING: break;
					}
					switch (checkServerMessages(lines[i]))
					{
						case RETURN: return;
						case CONTINUE: continue;
						case DO_NOTHING: break;
					}

				}
			}
			else if (++failCount >= 5) throw std::exception("Failed to recieve data from server.");
		}
	}
	catch (std::exception& e)
	{
		std::string ex = std::string(e.what());
		std::cout << ex << std::endl;
		disconnect();
		std::this_thread::sleep_for(std::chrono::seconds(10));
		main();
	}
	catch (...)
	{
		std::cout << "An exception occurred!" << std::endl;
		disconnect();
		std::this_thread::sleep_for(std::chrono::seconds(10));
		main();
	}
}

int main()
{
	Program* prog = new Program();
	prog->main();
	return 0;
}

File: IRC Bot Python/IrcBot.py
from socket import socket
from re import search as match_regex
from time import sleep


# Replacement for enum
class NextExecutionStep:
    def __init__(self):
        pass

    DO_NOTHING = 1
    CONTINUE = 2
    RETURN = 3


class IrcBot:
    real_name = ""
    user_name = ""
    host_name = ""
    nick_name = ""
    server = ""
    channel = ""
    port = 0
    irc = None
    names_to_op = []

    def __init__(self, real, user, host, nick, server_name, channel_name, server_port=6667):
        global real_name
        global user_name
        global host_name
        global nick_name
        global server
        global channel
        global port
        global names_to_op
        real_name = real
        user_name = user
        host_name = host
        nick_name = nick
        server = server_name
        channel = channel_name
        port = server_port
        names_to_op = ["Seanharrs", "SlayerSean",
                       "AusBotFSharp", "Aus_Bot_FSharp",
                       "AusBotCSharp", "Aus_Bot_CSharp",
                       "AusBotCPlusPlus", "Aus_Bot_CPlusPlus"]

    def main(self):
        encoding = "utf-8"

        def encode_message(text):
            return text.encode(encoding)

        def decode_message(text):
            return text.decode(encoding)

        def send_data(command):
            command = encode_message(command + "\n")
            print(command)
            irc.send(command)

        def connect():
            global irc
            irc = socket()
            irc.connect((server, port))
            send_data("NICK %s" % nick_name)
            # let the server register the nick command before sending the user command
            sleep(1)
            send_data("USER %s %s %s :%s" % (user_name, host_name, server, real_name))

        def disconnect():
            global irc
            irc.close()

        def join():
            send_data("JOIN %s" % channel)

        def try_op(name):
            for user in names_to_op:
                if user == name:
                    send_data("MODE %s" % (channel + " +o " + name))

        def check_server_messages(message):
            split_line = message.split(" ")
            if split_line[0] == "PING":
                send_data("PONG %s" % split_line[1])

            elif split_line[1] == "JOIN":
                name = match_regex(r":([^!]*)!", split_line[0]).group(1)
                if not (name is None or name == nick_name or name == user_name):
                    send_data("PRIVMSG %s" % (channel + " :Hello, " + name))
                    try_op(name)

            elif split_line[1] == "MODE":
                name = match_regex(r"([+-]o) (\S*)", message)
                if name is None:
                    return NextExecutionStep.CONTINUE
                command = name.group(1)
                name = name.group(2)

                if command == "-o":
                    try_op(name)

            else:
                return NextExecutionStep.DO_NOTHING

            return NextExecutionStep.CONTINUE

        def check_user_messages(server_message):
            user_message_begin_index = 3
            split_message = server_message.split(" ")

            if len(split_message) >= user_message_begin_index:
                if split_message[user_message_begin_index] == ":.botquit":
                    disconnect()
                    return NextExecutionStep.RETURN

                if split_message[user_message_begin_index] == ":.get":
                    print(server_message)
                    match = match_regex(":([^!]*)!.*?:.get ([\S ]*)", server_message)
                    name = match.group(1)
                    message = match.group(2).replace(" ", "_")
                    url = "https://en.wikipedia.org/wiki/" + message.lower()
                    send_data("PRIVMSG %s" % (channel + " :" + name + ": " + url))

            return NextExecutionStep.DO_NOTHING

        stream_buffer = ""
        count = 0
        joined = False
        try:
            connect()
            while True:
                # join after 4 messages as per IRC protocol
                if not joined and count >= 5:
                    joined = True
                    join()

                stream_buffer = stream_buffer + decode_message(irc.recv(1024))
                lines = stream_buffer.split("\n")
                print(lines)
                # the last element of the list (empty or un-finished message) becomes the buffer
                stream_buffer = lines.pop()
                print(stream_buffer)
                print(lines)
                for line in lines:
                    if "PING" not in line:
                        count += 1
                    # remove any trailing whitespace from the string (\t, \r, spaces, etc.)
                    line = line.rstrip()
                    result = check_server_messages(line)
                    if result == NextExecutionStep.RETURN:
                        return
                    if result == NextExecutionStep.CONTINUE:
                        continue

                    result = check_user_messages(line)
                    if result == NextExecutionStep.RETURN:
                        return
        except Exception as e:
            disconnect()
            print(str(e))
            sleep(10)
            return self.main()


if __name__ == "__main__":
    bot = IrcBot("Aus Bot Python v0", "AUS_Bot_Python", "AusBotHosting",
                 "AusBotPython", "irc.quakenet.org", "#ausbot_test")
    bot.main()
